api/v1alpha1/stresschaos_types.go

```diff
@@ -78,12 +77,18 @@ type StressChaosStatus struct {
 
 // StressInstance is an instance generates stresses
 type StressInstance struct {
-	// UID is the instance identifier
+	// CpuUID is the stress-ng identifier
 	// +optional
-	UID string `json:"uid"`
-	// StartTime specifies when the instance starts
+	CpuUID string `json:"cpuUid"`
+	// MemoryUID is the memStress identifier
 	// +optional
-	StartTime *metav1.Time `json:"startTime"`
+	MemoryUID string `json:"memoryUid"`
+	// CpuStartTime specifies when the stress-ng starts
+	// +optional
+	CpuStartTime *metav1.Time `json:"cpuStartTime"`
+	// MemoryStartTime specifies when the memStress starts
+	// +optional
+	MemoryStartTime *metav1.Time `json:"memoryStartTime"`
```

Comment:
Adding fields LGTM, but deleting/renaming not, that would bring break changes in our API. 

We could only update this field after multi-version support.
FYI @YangKeao @iguoyr 

Please consider still using `UID` and `StartTime` for CPU stress, and update the comments for them.

Reply:
We make breaking changes on JVMChaos in the last 2.1 release. Actually it's dirty as well. But there are not so widely used with JVMChaos, it might be acceptable. :cold_sweat:

Reply:
PTAL @FingerLeader

Reply:
It's necessary for canceling stressors, becasue `stress-ng` and `memStress` have different process UIDs. If we don't want to change the structure of `StressInstance`, we may need something like `stress-ng`'s "server" to manage them.

Reply:
@STRRL @FingerLeader I understand. The problem is that one stresschaos can create both cpu stressor and memory stressor at the same time, so we need two more fields to store the `PID` and `startedTime` for `memStress`.

One dirty solution would be adding a single parent process for these two processes:

```
bash -c "stress-ng & memStress & wait"
```

(I'm not sure whether it would work correctly, especially considering the `pause`, cgroup change and termination problem 😢 .

Reply:
But I agree with @STRRL , we shouldn't break the compatiblity. If the user is using the `StressChaos` and updating the Chaos Mesh, all existing `StressChaos` will break.

Reply:
Sorry @YangKeao , whether there are any problem that we parse API-struct like `StressInstance` to a struct like `StressInstanceInner` and use `StressInstanceInner` in apply & recover ?

Reply:
@FingerLeader Since type of `UID` is string ,we can store everything inside it. We can try marshual UID from,
```
type UID struct {
	CpuUID string `json:"cpuUid"`
	MemoryUID string `json:"memoryUid"`
}
```

Reply:
> 

No. It's not about the type. It's about whether we can handle the old resource.

Reply:
I just prefer to not change the field name of `UID` and `StartTime`, we could still append fields like `MemoryUID` and `MemroyStartTime`🤣

```go
type StressInstance struct {
	// UID is the stress-ng identifier
	// +optional
	UID string `json:"Uid"`
	// MemoryUID is the memStress identifier
	// +optional
	MemoryUID string `json:"memoryUid"`
	// StartTime specifies when the stress-ng starts
	// +optional
	StartTime *metav1.Time `json:"startTime"`
	// MemoryStartTime specifies when the memStress starts
	// +optional
	MemoryStartTime *metav1.Time `json:"memoryStartTime"`
}
```

Reply:
> @FingerLeader Since type of `UID` is string ,we can store everything inside it. We can try marshual UID from,
> 
> ```
> type UID struct {
> 	CpuUID string `json:"cpuUid"`
> 	MemoryUID string `json:"memoryUid"`
> }
> ```

Please DO NOT DO THAT anywhere, it's dangerous and would lose much information(the purpose of this field and developer's intuition), users and other developers would never know that could be a JSON or YAML or other serialized data. 😭 

A `string` type field could be often abused to storage any serialized data. It breaks a similar rule within database design: "multivalued column". Please treat `string` as other "primitive" type like `int` when you design some structure.

Reply:
> > @FingerLeader Since type of `UID` is string ,we can store everything inside it. We can try marshual UID from,
> > ```
> > type UID struct {
> > 	CpuUID string `json:"cpuUid"`
> > 	MemoryUID string `json:"memoryUid"`
> > }
> > ```
> 
> Please DO NOT DO THAT anywhere, it's dangerous and would lose much information(the purpose of this field and developer's intuition), users and other developers would never know that could be a JSON or YAML or other serialized data. 😭
> 
> A `string` type field could be often abused to storage any serialized data. It breaks a similar rule within database design: "multivalued column". Please treat `string` as other "primitive" type like `int` when you design some structure.

I think it is also dangerous to treat `string` as `int` if users do not have any preset on it. But if it seems dirty , we can use type definition:
```
type UIDString string

func (s *UIDString) ToUID() UID {
	pass
}

func (s *UIDString) FromUID(uid UID){
	pass
}

type UID struct {
	CpuUID string `json:"cpuUid"`
	MemoryUID string `json:"memoryUid"`
}

type StressInstance struct {
	// UID is the stress-ng identifier
	// +optional
	UID UIDString `json:"Uid"`
}
```

Reply:
I do not get your point @Andrewmatilde 🙈, we could not change the type of the field because of the compatibility.

Reply:
If we are using type definition ? or type alias ? , the yaml config definition of type will not be changed. 
Example : 
https://github.com/chaos-mesh/chaos-mesh/blob/master/api/v1alpha1/dnschaos_type.go#L66
```
type DNSChaosSpec struct {
	// Action defines the specific DNS chaos action.
	// Supported action: error, random
	// Default action: error
	// +kubebuilder:validation:Enum=error;random
	Action DNSChaosAction `json:"action"`
}
```
https://github.com/chaos-mesh/chaos-mesh/blob/master/helm/chaos-mesh/crds/chaos-mesh.org_dnschaos.yaml#L46
```
          spec:
            description: Spec defines the behavior of a pod chaos experiment
            properties:
              action:
                description: 'Action defines the specific DNS chaos action. Supported
                  action: error, random Default action: error'
                enum:
                - error
                - random
                type: string
``` 

& Because the client of user will send the json info to k8s api server first, the change will not break the compatibility.

Reply:
> we could still append fields like

@STRRL @Andrewmatilde I think just add fields is acceptable. I don't want to make the definition of StressInstance be as complicated as @Andrewmatilde said, and this solution eliminates our concerns about compatibility.

Reply:
@FingerLeader , I agree with you . The simple method may be better.